### DISCUSSION TITLE
Suggested improvements for version 2

### DISCUSSION THREAD

**Discussion by filiptammergard** [einride/eslint-plugin#31] [Ideas]
# Suggested improvements for version 2

## Currently extended configs and plugins

- [eslint-plugin-jest/recommended](https://www.npmjs.com/package/eslint-plugin-jest)
- [eslint-config-airbnb](https://www.npmjs.com/package/eslint-config-airbnb)
- [eslint-config-airbnb/hooks](https://www.npmjs.com/package/eslint-config-airbnb)
- [@typescript-eslint/eslint-plugin/recommended](https://www.npmjs.com/package/@typescript-eslint/eslint-plugin)
- [eslint-config-prettier/@typescript-eslint](https://www.npmjs.com/package/eslint-config-prettier)
- [eslint-plugin-prettier](https://www.npmjs.com/package/eslint-plugin-prettier)

## Suggested rule changes

We are disabling several rules. Some of them are disabled for good reasons and should remain disabled. Others are good rules that probably should be enabled, but are disabled for legacy reasons, to not have to do too many fixes when the eslint-config-airbnb was extended for example. Some of these rules should be enabled to ensure we're following good practices in our new projects.

A general suggestion is to avoid using warnings. Either error or disabled, to prevent warnings from piling up.

### [prefer-template](https://eslint.org/docs/rules/prefer-template)

Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/es6.js#L155).

```ts
// disallowed
const str = "Hello, " + name + "!";

// allowed
const str = `Hello, ${name}!`;
```

### [@typescript-eslint/no-explicit-any](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md)

Currently disabled. Is a warning in [@typescript-eslint/eslint-plugin/recommended](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/recommended.ts#L17). Suggestion is to make it an error.

```ts
// disallowed
const age: any = "seventeen";
```

### [@typescript-eslint/no-non-null-assertion](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-non-null-assertion.md)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [@typescript-eslint/eslint-plugin/recommended](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/recommended.ts#L18).

```ts
interface Foo {
  bar?: string;
}

const foo: Foo = getFoo();

// disallowed
const includesBaz = foo.bar!.includes('baz');

// allowed
const includesBaz = foo.bar?.includes('baz') ?? false;
```

### [jsx-a11y/anchor-is-valid](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react-a11y.js#L35).

```tsx
// disallowed
<a onClick={foo} />
<a href="#" onClick={foo} />
<a href="javascript:void(0)" onClick={foo} />
<a />
<a href={undefined} />
<a href={null} />
<a href="#" />
<a href="javascript:void(0)" />

// allowed
<a href={someValidPath} />
<a href={someValidPath} onClick={foo} />
```

### [no-else-return](https://eslint.org/docs/rules/no-else-return)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/best-practices.js#L94).

```ts
// disallowed
function foo() {
  if (x) {
    return y;
  } else {     
    return z;
  }
}

// allowed
function foo() {
  if (x) {
    return y;
  }

  return z;
}
```

### [no-restricted-globals](https://eslint.org/docs/rules/no-restricted-globals)

Currently disabled. `isNaN` and `isFinite` are broken but remain for legacy reasons. `Number.isNaN` and `Number.isFinite` should be used instead. Some browser globals commonly cause confusion and are not recommended to use without an explicit `window-` qualifier. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/variables.js#L19).

```ts
// disallowed
isNaN(x);
isFinite(x);
const a = status;

// allowed
Number.isNaN(x);
Number.isFinite(x);
const a = window.status;
```

### [no-shadow](https://eslint.org/docs/rules/no-shadow)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/variables.js#L34).

```ts
// disallowed
var a = 3;

function b() {
  var a = 10;
}
```

### [object-shorthand](https://eslint.org/docs/rules/object-shorthand)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/es6.js#L105).

```ts
// disallowed
const foo = {
  x: x,
  y: function () {},
};

// allowed
const foo = {
  x,
  y() {},
};
```

### [prefer-destructuring](https://eslint.org/docs/rules/prefer-destructuring)

Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/es6.js#L124).

```ts
const obj = { 
  firstName: "Filip", 
  lastName: "Tammergård"
};
const arr = [1, 2, 3];

// disallowed
const firstName = obj.firstName;
const lastName = obj.lastName;
const first = arr[0];
const second = arr[1];

// allowed
const { firstName, lastName } = obj;
const [first, second] = arr;
```

### [react/no-array-index-key](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react.js#L387).

```tsx
// disallowed
things.map((thing, index) => (
  <Hello key={index} />
));

// allowed
things.map((thing) => (
  <Hello key={thing.id} />
));
```

### [react/jsx-boolean-value](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-boolean-value.md)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react.js#L67).

```tsx
// disallowed
<Foo disabled={true} />

// allowed
<Foo disabled />
<Foo disabled={!!bar} />
```

### [react/jsx-key](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-key.md)

Currently warning. Suggestion is to make it an error.

```tsx
// disallowed
things.map((thing) => (
  <Hello />
));

// allowed
things.map((thing) => (
  <Hello key={thing.id} />
));
```

### [react/jsx-filename-extension](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-filename-extension.md)

Currently warning and `.js`, `.jsx` and `.tsx` are allowed JSX file extensions. Suggestion is to only allow `.tsx`, since we're exclusively using TypeScript in our React projects. Make it an error instead of warning.

### [@typescript-eslint/no-unused-vars](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars.md)

Currently disabled. Suggestion is to either remove it, which will make it a warning as in [@typescript-eslint/eslint-plugin/recommended](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/recommended.ts#L28), or to make it an error. If `"noUnusedLocals": true` in `tsconfig.json`, unused variables will be a ts warning. But we're using the default `"noUnusedLocals": false` in most projects. And better to use ESLint for linting than tsconfig IMO.

```ts
// disallowed
const a = "foo";

// a never used
```

### [consistent-return](https://eslint.org/docs/rules/consistent-return)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/best-practices.js#L27).

```ts
// disallowed
if (condition) {
  return true;
} else if (condition2) {
  return;
}

// allowed
if (condition) {
  return true;
} else if (condition2) {
  return false;
}

if (condition) {
  return;
} else if (condition2) {
  return;
}
```

### [default-case](https://eslint.org/docs/rules/default-case)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/best-practices.js#L35), where either an empty `default:` or a `// no default` comment satisfies the rule. Omitting `default` should be done explicitly to make the developer's intention clear.

```ts
// disallowed
switch (a) {
  case 1:
    doSomething();
    break;
}

// allowed
switch (a) {
  case 1:
    doSomething();
    break;

  default:
}

switch (a) {
  case 1:
    doSomething();
    break;

  // no default
}
```

### [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/style.js#L339), where [for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in), [for...of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of), [loop labels](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label) and [with statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with) are errors for [these reasons](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/style.js#L339).

### [react/jsx-curly-brace-presence](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-curly-brace-presence.md)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react.js#L431).

```tsx
// disallowed
<App>{"Hello world"}</App>;
<App prop={"Hello world"} attr={"foo"} />;

// allowed
<App>"Hello world"</App>;
<App prop="Hello world" attr="foo" />;
```

### [react/jsx-one-expression-per-line](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-one-expression-per-line.md)

Currently disabled. Suggestion is to remove disable, which will make it an error as in [eslint-config-airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react.js#L435).

```tsx
// disallowed
<App><Hello /><Hello2 /></App>

<App><Hello />
  <Hello2 />
</App>

<App>
  <Hello />
  <Hello2 /></App>

// allowed
<App>
  <Hello />
  <Hello2 />
</App>
```

### [react/jsx-wrap-multilines](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-wrap-multilines.md)

Currently disabled. Suggestion is to remove disabled, which will make it an error as in [eslint-config-airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react.js#L293).

```tsx
// disallowed
const hello = (<div>
  <p>Hello</p>
</div>);

// allowed
const hello = (
  <div>
    <p>Hello</p>
  </div>
);
```

### React 17

With React 17, [react/react-in-jsx-scope](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/react-in-jsx-scope.md) and [react/jsx-uses-react](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-uses-react.md) should be disabled (but enabled with React 16). Possible solutions:

- Let the rules remain in the plugin and disable them in all React 17 projects. (Bad solution IMO, since all our new projects are going to use React 17).
- Disable the rules in the plugin and enable them in the old projects where we want to upgrade to version 2 but continue using React 16.
- Instead of having a `default` config, we could split it up in one `recommended` config (with the rules disabled and one `react16` config (with the rules enabled).

### Development workflow

Some of these rules impair development workflow, when the app breaks if one variable isn't currently used for example when you're trying something out. Do we want to adopt some kind of workflow where some ESLint errors are OK during development, but pushes are rejected? Other ideas?

### Dependencies

- Bump dependencies to latest
- Should `prettier` be a dependency in a ESLint plugin, or rather a peerDependency? If we want `prettier` as a dependency, should we also include a default `.prettierrc` that can be extended/overwritten? Use fixed `prettier` version.

### [@typescript-eslint/explicit-function-return-type](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-function-return-type.md) and [@typescript-eslint/no-use-before-define](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/recommended.ts)

Currently disabled (and should be IMO). But the rules are not enabled in [@typescript-eslint/eslint-plugin/recommended](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/recommended.ts) that we're extending, so they can be removed from our plugin without any actual change.

### [no-undef](https://eslint.org/docs/rules/no-undef#disallow-undeclared-variables-no-undef)

Currently disabled. It's an error in [eslint-config-airbnb/base](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/variables.js#L40) but disabled in [@typescript-eslint/eslint-plugin/recommended](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/eslint-recommended.ts#L24) which comes later in the `extends` array. Therefore, we could remove it from our plugin without any actual change.

### [import/resolver](https://github.com/benmosher/eslint-plugin-import#importextensions)

Currently, `.js`, `.jsx`, `.ts` and `.tsx` will be parsed as modules and inspected for exports. Suggestion is to change to `.ts` and `.tsx`, since we're exclusively using TypeScript in our React projects.

### "jest/globals": true

`"jest/globals": true` is included through [eslint-plugin-jest/recommended](https://github.com/jest-community/eslint-plugin-jest/blob/main/src/index.ts#L63) and can be removed without any actual change.

**Comment by filiptammergard** [comment]
Comments and additions:

- react/jsx-filename-extension: Keep as is to keep supporting `.js` and `jsx`.
- import/resolver: Keep as is to keep supporting `.js` and `jsx`.
- Remove disable of `react/jsx-props-no-spreading` rule, which will make it an error as in [eslint-config-airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react.js#L495).
- Enforce import organization.
- Disable React 17 rules, and enable them in React 16 projects.

**Comment by ericwenn** [comment]
Amazing job compiling this list! Agree with almost every suggestion here, and glad to see many of these rules being enabled again.

Will take a closer look next week!
